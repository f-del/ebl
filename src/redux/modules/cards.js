import update from "immutability-helper";

const CREATE = "CARD/CREATE";
const START = "CARD/START";

const card = title => ({
  Title: title,
  Status: cardStatus.Todo,
  Type: cardType.Task
});

export const cardStatus = {
  Todo: "TODO",
  Inprogress: "INPROGRESS"
};

export const cardType = {
  Task: "TASK"
};

export const createCardSuccess = (id, card) => ({
  type: CREATE,
  payload: { Id: id, ...card }
});

export const createCard = title => {
  return (dispatch, getState, { api }) => {
    const newCard = card(title);
    return api.Tasks.Post(newCard).then(res => {
      console.log("post done,  new Id =" + res.Id);
      dispatch(createCardSuccess(res.Id, newCard));
    });
  };
};

export const startCard = id => {
  if (id === undefined) throw new Error("Id arg is mandatory");
  return {
    type: START,
    payload: {
      Id: id
    }
  };
};

const initialState = {
  list: []
};

export default function(state = initialState, action) {
  console.debug(action);
  switch (action.type) {
    case CREATE:
      return { ...state, list: [...state.list, action.payload] };
    case START:
      let idx = state.list.findIndex(c => c.Id === action.payload.Id);

      var mergeObject = {};
      mergeObject[idx] = { $merge: { Status: cardStatus.Inprogress } };

      return { ...state, list: update(state.list, mergeObject) };

    default:
      return state;
  }
}

export function getAllCards(state) {
  if (state === undefined) throw new Error("State arg is mandatory");
  return state.cards !== undefined ? state.cards.list : [];
}

export function getAllCardsInProgess(state) {
  return getAllCards(state).filter(isCardStatusInProgess);
}
export function getAllCardsTodo(state) {
  return getAllCards(state).filter(isCardStatusToDo);
}

/* Functionnal helper */
function isCardStatusInProgess(card) {
  return card.Status === cardStatus.Inprogress;
}
function isCardStatusToDo(card) {
  return card.Status === cardStatus.Todo;
}
