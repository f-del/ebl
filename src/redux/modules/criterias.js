export const criteriaType = {
  BASIC: "BASIC",
  DEV: "DEV"
};

export function create(id, value) {
  return { Id: id, Value: value };
}

const initialState = {
  list: {
    BASIC: [
      {
        Id: "IdCritBas1",
        Value: false
      }
    ],
    DEV: [
      {
        Id: "IdCritDev1",
        Text: "Development",
        Value: false
      },
      {
        Id: "IdCritDev2",
        Text: "Unit testing",
        Value: false
      }
    ]
  }
};

export default function(state = initialState, action) {
  return state;
}

export function getCriteria(state, type) {
  if (type === undefined) throw new Error("Type argument is mandatory");
  return getState(state).list[type] || [];
}
function getState(state) {
  return state.criterias || {};
}
