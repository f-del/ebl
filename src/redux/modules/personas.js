/***
 *    ██████╗ ███████╗██████╗ ██╗   ██╗ ██████╗███████╗██████╗ ███████╗
 *    ██╔══██╗██╔════╝██╔══██╗██║   ██║██╔════╝██╔════╝██╔══██╗██╔════╝
 *    ██████╔╝█████╗  ██║  ██║██║   ██║██║     █████╗  ██████╔╝███████╗
 *    ██╔══██╗██╔══╝  ██║  ██║██║   ██║██║     ██╔══╝  ██╔══██╗╚════██║
 *    ██║  ██║███████╗██████╔╝╚██████╔╝╚██████╗███████╗██║  ██║███████║
 *    ╚═╝  ╚═╝╚══════╝╚═════╝  ╚═════╝  ╚═════╝╚══════╝╚═╝  ╚═╝╚══════╝
 *
 */

const initialState = {
  list: {
    e6wTJIBsAw5puwwrNEsR: {
      Name: "Feature Team",
      Details: "Meta persona, representing all collaborators working with Paul",
      Needs: [
        "To understand WHAT the Product Owner wants to be able to design collaborativly the HOW",
        "To manage easly the list of things to do to finish the most quickly the User Story"
      ],
      Picture: ""
    },
    uOCF2GLFiJl6fvxdlq4b: {
      Details: "Product Owner for a new product, work in a startup",
      Name: "Paul Deligny",
      Needs: [
        "To have a clear understanding about the product vision & objectives",
        "To write real User Stories, ie have a Customer Driven thinking",
        "To share at the right time and with the rigth level of information the User Stories with his team",
        "To have a clear vision on the job he'll have to do AND the job already or currently done"
      ],
      Picture: ""
    }
  }
};

export default function(state = initialState, action) {
  return state;
}

/***
 *    ███████╗███████╗██╗     ███████╗ ██████╗████████╗ ██████╗ ██████╗ ███████╗
 *    ██╔════╝██╔════╝██║     ██╔════╝██╔════╝╚══██╔══╝██╔═══██╗██╔══██╗██╔════╝
 *    ███████╗█████╗  ██║     █████╗  ██║        ██║   ██║   ██║██████╔╝███████╗
 *    ╚════██║██╔══╝  ██║     ██╔══╝  ██║        ██║   ██║   ██║██╔══██╗╚════██║
 *    ███████║███████╗███████╗███████╗╚██████╗   ██║   ╚██████╔╝██║  ██║███████║
 *    ╚══════╝╚══════╝╚══════╝╚══════╝ ╚═════╝   ╚═╝    ╚═════╝ ╚═╝  ╚═╝╚══════╝
 *
 */
export function getAllPersonas(state) {
  return mapToArrayPersona(getPersonasList(state)) || [];
}

function getPersonasList(state) {
  return getPersonas(validateState(state)).list || {};
}

/* Functionnal helper */
function validateState(state) {
  if (state === undefined) throw new Error("State arg is mandatory");
  return state;
}
function getPersonas(state) {
  return state.personas || {};
}

function mapToArrayPersona(list) {
  return Array.from(Object.keys(list), key => ({
    ...getPersonaId(key),
    ...list[key]
  }));
}
function getPersonaId(key) {
  return {
    Id: key
  };
}
