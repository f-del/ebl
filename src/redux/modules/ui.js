const SELECT_HYPOTHESIS = "UI/SELECT/HYPOTHESIS";
const SELECT_USERSTORY = "UI/SELECT/USER-STORY";

export const selectHypothesis = hypothesis => ({
  type: SELECT_HYPOTHESIS,
  payload: {
    Id: hypothesis.Id
  }
});

export const selectUserStory = userstory => ({
  type: SELECT_USERSTORY,
  payload: {
    Id: userstory.Id
  }
});

const initialState = {
  currentHypothesis: undefined,
  currentUserStory: undefined
};

export default function uiReducer(state = initialState, action) {
  switch (action.type) {
    case SELECT_HYPOTHESIS: {
      // currently Reducer handle Ui logic -> deselect current US on Hypothesis change
      return {
        ...state,
        currentHypothesis: action.payload.Id,
        currentUserStory: undefined
      };
    }
    case SELECT_USERSTORY: {
      return { ...state, currentUserStory: action.payload.Id };
    }

    default:
      return state;
  }
}

export function getHypothesisSelected(state) {
  return state.ui.currentHypothesis;
}
export function getUserStorySelected(state) {
  return state.ui.currentUserStory;
}
