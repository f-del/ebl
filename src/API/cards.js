import { cardStatus } from "../redux/modules/cards";

const mapping = card => {
  card = card._firestore !== undefined ? card.data() : card;
  return {
    ...card,
    Status:
      card.Status === "INPROGRESS" ? cardStatus.INPROGRESS : cardStatus.TODO
  };
};

function post(db) {
  return card => {
    return new Promise((resolve, reject) => {
      db.collection("cards")
        .add(card)
        .then(ref => {
          console.log("Document successfully written!");
          resolve({ id: ref.id });
        })
        .catch(error => {
          console.error("Error writing document: ", error);
          //dispatch(errorSave())
          reject(error);
        });
    });
  };
}

function get(db) {
  return type => {
    return new Promise((resolve, reject) => {
      db.collection("cards")
        .where("Type", "==", type)
        .get()
        .then(function(querySnapshot) {
          let cards = [];

          querySnapshot.forEach(function(doc) {
            let env = mapping(doc);

            cards.push(env);
          });

          resolve(cards);
        });
    });
  };
}

export default db => {
  return { Post: post(db), Get: get(db) };
};
