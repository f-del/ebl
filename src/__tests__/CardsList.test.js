import React from "react";
import { configure, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";

import CardListByStatus from "../containers/CardsListByStatus";

import { cardStatus } from "../redux/modules/cards";
import * as cardsSelector from "../redux/modules/cards";

configure({ adapter: new Adapter() });
let store;

describe("Containers tests", () => {
  describe("CardsListByStatus", () => {
    beforeEach(() => {
      store = configureMockStore();
      jest.resetAllMocks();
    });
    test("CardsListByStatus with bad symbol status parameter", () => {
      const falseStatus = Object.freeze({
        TODO: Symbol("FALSETODO")
      });
      const wrapper = () => {
        const wrapper = mount(
          <Provider store={store({})}>
            <CardListByStatus status={falseStatus.TODO} />
          </Provider>
        );
        wrapper.unmount();
      };

      expect(wrapper).toThrowError(
        "Status not recognized, should come from cardStatus"
      );
    });

    test("CardsListByStatus with todo", () => {
      cardsSelector.getAllCardsTodo = jest.fn(state => {
        return [];
      });

      const wrapper = mount(
        <Provider store={store({})}>
          <CardListByStatus status={cardStatus.TODO} />
        </Provider>
      );

      expect(cardsSelector.getAllCardsTodo.mock.calls.length).toBe(1);
      wrapper.unmount();
    });

    test("CardsListByStatus with Inprogress", () => {
      cardsSelector.getAllCardsInProgess = jest.fn(state => {
        return [];
      });

      const wrapper = mount(
        <Provider store={store({})}>
          <CardListByStatus status={cardStatus.INPROGRESS} />
        </Provider>
      );

      expect(cardsSelector.getAllCardsInProgess.mock.calls.length).toBe(1);
      wrapper.unmount();
    });

    test("CardsListByStatus with DONE", () => {
      cardsSelector.getAllCardsDone = jest.fn(state => {
        return [];
      });

      const wrapper = mount(
        <Provider store={store({})}>
          <CardListByStatus status={cardStatus.DONE} />
        </Provider>
      );

      expect(cardsSelector.getAllCardsDone.mock.calls.length).toBe(1);
      wrapper.unmount();
    });
  });
});

describe("Componant tests", () => {});
