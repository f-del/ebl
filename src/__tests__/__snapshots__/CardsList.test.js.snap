// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Containers tests CardsListByStatus CardsListByStatus with DONE 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(CardsList)
    filterBy={
      Object {
        "needsIndex": 2,
        "persona": "1",
      }
    }
    status={Symbol(DONE)}
    type="TASK"
  >
    <CardsList
      filterBy={
        Object {
          "needsIndex": 2,
          "persona": "1",
        }
      }
      list={Array []}
      status={Symbol(DONE)}
      type="TASK"
    />
  </Connect(CardsList)>
</Provider>
`;

exports[`Containers tests CardsListByStatus CardsListByStatus with Inprogress 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(CardsList)
    status={Symbol(INPROGRESS)}
    type="USER-STORY"
  >
    <CardsList
      list={Array []}
      status={Symbol(INPROGRESS)}
      type="USER-STORY"
    />
  </Connect(CardsList)>
</Provider>
`;

exports[`Containers tests CardsListByStatus CardsListByStatus with todo 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(CardsList)
    status={Symbol(TODO)}
    type="HYPOTHESIS"
  >
    <CardsList
      list={Array []}
      status={Symbol(TODO)}
      type="HYPOTHESIS"
    />
  </Connect(CardsList)>
</Provider>
`;
